Resize canvas when changing figure size

dstansby opened the pull request:
When the figure size is resized, as well as the window being manually resized the canvas needs to be resized too. Usually the canvas is resized automatically to the window at draw time, but this handles the case where the figure size is changed more than once before drawing (ie. #15040).

This might not be the right thing to do though, maybe others know better?

Fixes #15040

tacaswell commented:
What is the mechanism by which this breaks?

It is also possible that this "just works" on master, @anntzer recently put in some PRs to tweak how we handle the window sizing.

dstansby commented:
See the example in https://github.com/matplotlib/matplotlib/issues/15040, which shows how it breaks, even on master.

tacaswell commented:
I mean, why does setting the size twice confuse it?

dstansby commented:
Currently the resize code in `Qt5` is:
```python
    def resize(self, width, height):
        # these are Qt methods so they return sizes in 'virtual' pixels
        # so we do not need to worry about dpi scaling here.
        extra_width = self.window.width() - self.canvas.width()
        extra_height = self.window.height() - self.canvas.height()
        self.window.resize(width + extra_width, height + extra_height)
```
After one resize the window gets resized, but the canvas doesn't. This means when it comes to the second resize, `self.canvas.width()/height()` don't reflect the previous resize, so `extra_width/height` are not computed correctly.

My guess as to why it works fine in the usual case with one resize and a draw is because at draw time the canvas is automatically resized to fit the window.

dstansby commented:
I've just pushed a test that fails without the change in this PR.

anntzer commented:
I'm not totally convinced this is the correct fix?  For example, perhaps set_size_inches should be the one doing the canvas resize, e.g. with
```
        if forward:
            canvas = getattr(self, 'canvas')
            if canvas is not None:
                dpi_ratio = getattr(canvas, '_dpi_ratio', 1)

                canvas.resize(...)  # <--- HERE

                manager = getattr(canvas, 'manager', None)
                if manager is not None:
                    manager.resize(*(size * self.dpi / dpi_ratio).astype(int))
```
otherwise, I would guess the other backends also need updating:
```
    # GTK3
    def resize(self, width, height):
        'set the canvas size in pixels'
        #_, _, cw, ch = self.canvas.allocation
        #_, _, ww, wh = self.window.allocation
        #self.window.resize (width-cw+ww, height-ch+wh)
        self.window.resize(width, height)

    # Tk
    def resize(self, width, height):
        self.canvas._tkcanvas.master.geometry("%dx%d" % (width, height))

        if self.toolbar is not None:
            self.toolbar.configure(width=width)

    # wx
    def resize(self, width, height):
        'Set the canvas size in pixels'
        self.canvas.SetInitialSize(wx.Size(width, height))
        self.window.GetSizer().Fit(self.window)
```
which appear to all suffer from the same issue?  (FigureManagerMac doesn't even seem to define resize()?)

dstansby commented:
Yes, I think you're right that `set_size_inches` should handle it, and what you proposed is correct, ie.

1. Change the figure (really `canvas`) size (currently missing)
2. If there's a window manager, change the window size to match (currently present)

dstansby commented:
Is there a way to parameterize the test with different backends?

timhoffm commented:
> Is there a way to parameterize the test with different backends?

Not sure. You can always generate the functions dynamically in a for loop. I think something like this should work:

~~~
for backend in ['Qt5Agg', 'TkAgg']:
    @pytest.mark.backend(backend)
    def _test_x():
        ...
    
    globals()[f'test_x_{backend}'] = _test_x

anntzer commented:
They basically need to run in different subprocesses (because a given process can only ever use a single interactive backend) -- see tests_backends_interactive.py.  (Probably stashing all the tests in a single process invocation, as is done right now, is better for performance.)

tacaswell commented:
With https://github.com/matplotlib/matplotlib/pull/14703 merged do we still need this?

dstansby commented:
Can take a look soon-ish, but I think it's worth putting the test in even if it does pass on current master.

dstansby commented:
Yep, this fix is still needed even with https://github.com/matplotlib/matplotlib/pull/14703

timhoffm commented on lib/matplotlib/tests/test_backend_qt.py:
```diff
@@ -272,3 +272,19 @@ def test_figureoptions():
+    window = fig.canvas.manager.window
+
+    fig.set_size_inches(2, 2)
+    old_width = window.width()
```
In my test of this code, the example window is always 640x480 no matter what I pass to `set_size_inches`. Maybe explicitly test for 200 here (or `rcParams['figure.dpi'] * 2` if you want to be very explicit).

dstansby commented:
I've changed my mind and think that it's reasonable for the backend `resize` methods to also be responsible for resizing the canvas. I'm going to leave this as is as a `Qt5` fix, and open new bugs for each other backend that has `set_size_inches()` issues.

I've also added an explict check for figure width as @timhoffm suggests.
